go/obscuronode/enclave/sql/sql.go

```diff
@@ -0,0 +1,224 @@
+	stmt := `create table if not exists kv (key text primary key, value blob); delete from kv;`
+	if _, err := m.db.Exec(stmt); err != nil {
+		return fmt.Errorf("failed to initialise sql eth db - %w", err)
+	}
+
+	return nil
+}
+
+func (m *sqlEthDatabase) Has(key []byte) (bool, error) {
+	pFind, err := m.getFindPrepStmt()
+	if err != nil {
+		return false, err
+	}
+	defer pFind.Close()
+
+	err = pFind.QueryRow(key).Scan()
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (m *sqlEthDatabase) Get(key []byte) ([]byte, error) {
+	var res []byte
+	pFind, err := m.getFindPrepStmt()
+	if err != nil {
+		return []byte{}, err
+	}
+	defer pFind.Close()
+
+	err = pFind.QueryRow(key).Scan(&res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+func (m *sqlEthDatabase) Put(key []byte, value []byte) error {
+	pIns, err := m.getInsertPrepStmt()
+	if err != nil {
+		return err
+	}
+	defer pIns.Close()
+
+	_, err = pIns.Exec(key, value)
+	return err
+}
+
+func (m *sqlEthDatabase) Delete(key []byte) error {
+	pDel, err := m.getDeletePrepStmt()
+	if err != nil {
+		return err
+	}
+	defer pDel.Close()
+
+	_, err = pDel.Exec(key)
+	return err
+}
+
+func (m *sqlEthDatabase) Close() error {
+	if err := m.db.Close(); err != nil {
+		return fmt.Errorf("failed to close sql db - %w", err)
+	}
+	return nil
+}
+
+func (m *sqlEthDatabase) NewBatch() ethdb.Batch {
+	log.Trace("SQL :: New batch")
+	return &sqlBatch{
+		db: m,
+	}
+}
+
+func (m *sqlEthDatabase) NewIterator(prefix []byte, start []byte) ethdb.Iterator {
+	pFindLike, err := m.getIteratorPrepStmt()
+	defer func() { _ = pFindLike.Close() }()
+	if err != nil {
+		return &iterator{
+			err: fmt.Errorf("failed to get prepared SQL stmt, iter will be empty, %w", err),
+		}
+	}
+	// todo: make sure we're stringifying that prefix correctly
+	pr := string(prefix)
+	st := string(append(prefix, start...))
+	rows, err := pFindLike.Query(pr+"%", st) //nolint:sqlclosecheck
+	defer func() { _ = pFindLike.Close() }()
+	if err != nil {
+		return &iterator{
+			err: fmt.Errorf("failed to get rows, iter will be empty, %w", err),
+		}
+	}
+	if err = rows.Err(); rows.Err() != nil {
+		return &iterator{
+			err: fmt.Errorf("failed to get rows, iter will be empty, %w", err),
+		}
+	}
+	return &iterator{
+		rows: rows,
+	}
+}
+
+func (m *sqlEthDatabase) Stat(property string) (string, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *sqlEthDatabase) Compact(start []byte, limit []byte) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *sqlEthDatabase) getPrepStmt(query string) (*sql.Stmt, error) {
+	prep, err := m.db.Prepare(query)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare statement `%s` - %w", query, err)
+	}
+	return prep, nil
+}
+
+func (m *sqlEthDatabase) getFindPrepStmt() (*sql.Stmt, error) {
+	return m.getPrepStmt(`select kv.value from kv where kv.key = ?;`)
```

Comment:
can the prepared statements be cached? It's what I remember from my pre-hibernate days :)

Reply:
Yeah I had them as cached properties on the struct at first and I think there's definitely an optimisation there. But then I saw some docs that scared me (about it being fragile if your connection was interrupted).

But I think you're right, I think I should be able to cache them, I think this is already fragile if the connection is interrupted but I think getting the reconnect logic right might be beyond the scope for now, if reload from persistence works then users can restart their nodes in the (hopefully) unlikely event of disconnect.

Reply:
have a look if there's any library out there. I would find it very weird if these were not solved problems.